File: projects/assistant-email/src/email_assistant/notify/formatter.py
"""Deterministic summary formatting.

The LLM writes prose; this module writes the *skeleton* — the block the user
can trust to always look the same and to reflect ground-truth counts.

Kept pure (no I/O): easy to unit-test, easy to reuse as an email body, a
Slack message, or a log line.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass

from email_assistant.domain.email import Email, Priority

_SECTIONS: tuple[tuple[Priority, str], ...] = (
    (Priority.URGENT, "🚨 URGENT"),
    (Priority.IMPORTANT, "⭐ IMPORTANT"),
    (Priority.LOW, "📋 LOW"),
)
_MAX_ITEMS_PER_SECTION = 5


@dataclass(frozen=True, slots=True)
class Summary:
    """Structured summary — good for tests, logging, and templating."""

    total: int
    by_priority: dict[Priority, list[Email]]

    @property
    def urgent_count(self) -> int:
        return len(self.by_priority.get(Priority.URGENT, []))

    @property
    def has_urgent(self) -> bool:
        return self.urgent_count > 0


@dataclass(frozen=True, slots=True)
class SummaryFormatter:
    """Builds a :class:`Summary` and renders it as Markdown."""

    max_items_per_section: int = _MAX_ITEMS_PER_SECTION

    # ------------------------------------------------------------------ build

    def summarize(self, emails: Iterable[Email]) -> Summary:
        buckets: dict[Priority, list[Email]] = {p: [] for p in Priority}
        total = 0
        for e in emails:
            total += 1
            priority = e.priority or Priority.LOW
            buckets[priority].append(e)
        return Summary(total=total, by_priority=buckets)

    # ------------------------------------------------------------------ render

    def render(self, summary: Summary) -> str:
        lines: list[str] = [
            f"# Email Summary — {summary.total} emails "
            f"({summary.urgent_count} urgent)",
            "",
        ]
        for priority, header in _SECTIONS:
            bucket = summary.by_priority.get(priority, [])
            if not bucket:
                continue
            lines.append(f"## {header} ({len(bucket)})")
            for e in bucket[: self.max_items_per_section]:
                lines.append(f"- **{e.subject}** — {e.sender} ({e.date})")
            if len(bucket) > self.max_items_per_section:
                extra = len(bucket) - self.max_items_per_section
                lines.append(f"- …and {extra} more")
            lines.append("")
        return "\n".join(lines).rstrip() + "\n"

    def summarize_and_render(self, emails: Iterable[Email]) -> tuple[Summary, str]:
        summary = self.summarize(emails)
        return summary, self.render(summary)

File: projects/assistant-email-go/notify/formatter.go
// Package notify renders the deterministic summary skeleton.
//
// The LLM writes prose; this package writes the block the user can trust to
// always look the same and to reflect ground-truth counts. Kept pure (no I/O).
package notify

import (
	"fmt"
	"strings"

	"github.com/m1981/temporal-go-agent-sdk/projects/assistant-email-go/domain"
)

const defaultMaxItemsPerSection = 5

var sections = []struct {
	priority domain.Priority
	header   string
}{
	{domain.PriorityUrgent, "🚨 URGENT"},
	{domain.PriorityImportant, "⭐ IMPORTANT"},
	{domain.PriorityLow, "📋 LOW"},
}

// Summary is the structured digest — good for tests, logging, and templating.
type Summary struct {
	Total      int
	ByPriority map[domain.Priority][]domain.Email
}

func (s Summary) UrgentCount() int { return len(s.ByPriority[domain.PriorityUrgent]) }
func (s Summary) HasUrgent() bool  { return s.UrgentCount() > 0 }

// Formatter builds a Summary and renders it as Markdown.
// The zero value uses defaultMaxItemsPerSection.
type Formatter struct {
	MaxItemsPerSection int
}

func (f Formatter) maxItems() int {
	if f.MaxItemsPerSection > 0 {
		return f.MaxItemsPerSection
	}
	return defaultMaxItemsPerSection
}

// Summarize buckets emails by priority. Unclassified emails count as LOW.
func (f Formatter) Summarize(emails []domain.Email) Summary {
	buckets := map[domain.Priority][]domain.Email{
		domain.PriorityUrgent:    {},
		domain.PriorityImportant: {},
		domain.PriorityLow:       {},
	}
	for _, e := range emails {
		p := e.Priority
		if p == "" {
			p = domain.PriorityLow
		}
		buckets[p] = append(buckets[p], e)
	}
	return Summary{Total: len(emails), ByPriority: buckets}
}

// Render produces the stable Markdown block.
func (f Formatter) Render(s Summary) string {
	var b strings.Builder
	fmt.Fprintf(&b, "# Email Summary — %d emails (%d urgent)\n\n", s.Total, s.UrgentCount())
	for _, sec := range sections {
		bucket := s.ByPriority[sec.priority]
		if len(bucket) == 0 {
			continue
		}
		fmt.Fprintf(&b, "## %s (%d)\n", sec.header, len(bucket))
		shown := bucket
		if len(shown) > f.maxItems() {
			shown = shown[:f.maxItems()]
		}
		for _, e := range shown {
			fmt.Fprintf(&b, "- **%s** — %s (%s)\n", e.Subject, e.Sender, e.Date)
		}
		if extra := len(bucket) - f.maxItems(); extra > 0 {
			fmt.Fprintf(&b, "- …and %d more\n", extra)
		}
		b.WriteString("\n")
	}
	return strings.TrimRight(b.String(), "\n") + "\n"
}

// SummarizeAndRender is the one-call form used by the pipeline.
func (f Formatter) SummarizeAndRender(emails []domain.Email) (Summary, string) {
	s := f.Summarize(emails)
	return s, f.Render(s)
}
